Scripts/Source/minai_Sex.psc

```diff
@@ -1122,3 +1106,165 @@ string Function GetFactionsForActor(actor akTarget)
+    sslThreadController controller = slf.GetController(ThreadID)
+  
+    if (controller.Stage==1) 
+      LoadSexlabDescriptions()
+    EndIf
+    
+    actors = slf.GetController(ThreadID).Positions
+    sceneId = controller.Animation.FetchStage(controller.Stage)[0]
+  else
+    actors = OThread.GetActors(ThreadID)
+    sceneId = OThread.GetScene(ThreadID)
+    speed = OThread.GetSpeed(ThreadID)
+  endif
+
+  string actorsString = JoinActorArray(actors)
+
+  if(fallback == "")
+    fallback = actorsString + " have sex."
+  endif
+
+  AIAgentFunctions.logMessage("command@ExtCmdUpdateThreadsTable@{ \"type\": \""+type+"\", \"framework\": \""+framework+"\", \"threadId\": "+ThreadID+", \"actors\": \""+actorsString+"\", \"scene\": \""+sceneId+"\", \"speed\": "+speed+", \"fallback\": \""+type+"\"}@", "updateThreadsDB")
+endfunction
+
+function sexTalkClimax(actor speaker, bool chastity = false)
+  if(chastity)
+    SexTalk(speaker, "chatnf_sexclimaxchastity")
+  else
+    SexTalk(speaker, "chatnf_sexclimax")
+  endif
+endfunction
+
+function sexTalkSceneChage(actor speaker)
+  SexTalk(speaker, "chatnf_sexscenechange")
+endfunction
+
+function sexTalkSpeedChange(actor speaker)
+  SexTalk(speaker, "chatnf_sexspeedchange")
+endfunction
+
+function sexTalkOnEnd(actor speaker)
+  SexTalk(speaker, "chatnf_sexend")
+endfunction
+
+function sexTalkCollision(actor speaker, string promptToSay)
+  SexTalk(speaker, "chatnf_vr_1")
+endfunction
+
+; speaker is who actually will say llm lines, can be none if scene is player only
+; chatType different AIFF custom chat topics see php files
+Function SexTalk(actor speaker, string chatType)
+  if !bHasAIFF || !speaker
+    return
+  EndIf
+
+  ; Throttle on how often we should dirty talk incase people are switching animations
+  float currentTime = Utility.GetCurrentRealTime()
+  if currentTime - lastDirtyTalk > 5
+    lastDirtyTalk = currentTime
+    string speakerName = speaker.GetDisplayName()
+    Main.Debug("DirtyTalk() => " + speakerName + ": " + chatType)
+    Main.RequestLLMResponseNPC("", "", speakerName, chatType)
+  else
+    Main.Debug("DirtyTalk - THROTTLED")
+  EndIf
+EndFunction
+
+actor function getRandomActor(actor[] actors)
+  int index = Utility.RandomInt(0, actors.length - 1)
+
+  return actors[index]
+endfunction
+
+; if it's player only scene it will return none
+; assume that actors are only actors from scene
+actor Function GetWeightedRandomActorToSpeak(actor[] actors)
+  actor[] maleActors = PapyrusUtil.ActorArray(0)
+  actor[] femaleActors = PapyrusUtil.ActorArray(0)
+  int count = actors.Length
+  while count > 0
+    count -= 1
+    actor currActor = actors[count]
+    int currSex = currActor.GetActorBase().GetSex()
+    bool isMuted = false
+
+    ; if ostim actor is muted
+    if(bHasOstim)
+      if(OActor.IsMuted(currActor))
+        isMuted = true
+      endif
+    endif
+
+    if(currActor == PlayerRef || isMuted)
+      ; player doesn't participate in llm talking :)
```

Comment:
Text from "the player" which has an NPC respond is a valid usecase, right?

Reply:
Yeah when player is asking something directly llm will answer as usual. These speaker when npc want to say something on their own(comment scene change, speed change, climax, collision or just random comment) which doesn't require player's direct input